ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -115,22 +115,27 @@ forkSyncStateOnTipPointChange registry menv = do
 implAddTxs :: forall m blk hdr. (MonadSTM m, Ord (GenTxId blk), ApplyTx blk)
            => MempoolEnv m blk hdr
            -> [(GenTxId blk, GenTx blk)]
-           -> m [(GenTx blk, ApplyTxErr blk)]
+           -> m [(GenTx blk, Maybe (ApplyTxErr blk))]
 implAddTxs mpEnv@MempoolEnv{mpEnvStateVar, mpEnvLedgerCfg} txs =
     atomically $ do
       ValidationResult {
         vrBefore,
         vrValid,
+        vrNewValid,
         vrInvalid
       } <- validateNew <$> validateIS mpEnv
       writeTVar mpEnvStateVar IS { isTxs   = vrValid
                                  , isTip   = vrBefore
                                  }
-      return vrInvalid
+      pure $ (vrInvalidMaybeErr vrInvalid) ++ vrNewValid
```

Comment:
@mrBliss Do you think this is too inefficient? The alternative would probably be maintaining an additional field in the `ValidationResult` which contains both the newly added and all invalid transactions: `vrNewValidAndAllInvalid :: [(GenTx blk, Maybe (ApplyTxErr blk))]` (quite a long name, but I'm not sure what else I'd call this field in order to get the point across) in addition to the currently existing fields, `vrValid` and `vrInvalid`.

Reply:
I don't think this is inefficient at all, the concatenation is lazy, so this is fine :+1: